src/automaton/builder.rs

```diff
@@ -0,0 +1,343 @@
+    pub fn new() -> Self
+    where
+        I: Default,
+    {
+        Self::default()
+    }
+
+    /// Construct an automaton builder from a list of patterns, given by lists of
+    /// constraints.
+    ///
+    /// Use `I::default()` as the indexing scheme.
+    pub fn from_constraints(patterns: Vec<Vec<C>>) -> Self
+    where
+        I: Default,
+    {
+        Self {
+            patterns,
+            matcher: ConstraintAutomaton::new(),
+        }
+    }
+
+    /// Construct an automaton builder from a list of patterns with a custom
+    /// indexing scheme.
+    pub fn from_constraints_with_index_scheme(patterns: Vec<Vec<C>>, host_indexing: I) -> Self {
+        Self {
+            patterns,
+            matcher: ConstraintAutomaton::with_indexing_scheme(host_indexing),
+        }
+    }
+
+    /// Add a pattern to the automaton builder.
+    pub fn add_pattern(&mut self, pattern: Vec<C>) {
+        self.patterns.push(pattern);
+    }
+}
+
+type Automaton<K, P, I> = ConstraintAutomaton<Constraint<K, P>, I>;
+
+impl<K, P, I> AutomatonBuilder<Constraint<K, P>, I>
+where
+    Constraint<K, P>: Eq + Clone + ToConstraintsTree,
+{
+    /// Construct the automaton.
+    ///
+    /// The returned automaton will be able to match `self.patterns` and will
+    /// respect the automaton invariants: at every state and for any input, the transitions are mutually exclusive
+    /// or an epsilon transition.
+    ///
+    /// The `make_det` predicate specifies the heuristic used to determine whether
+    /// to turn a state into a deterministic one.
+    pub fn build(
+        mut self,
+        mut make_det: impl for<'c> FnMut(&[&'c Constraint<K, P>]) -> bool,
+    ) -> (Automaton<K, P, I>, Vec<PatternID>) {
+        // Construct a prefix tree by adding all constraints non-deterministically
+        let pattern_ids = mem::take(&mut self.patterns)
+            .into_iter()
+            .enumerate()
+            .map(|(id, constraints)| {
+                let id = PatternID(id);
+                self.matcher.add_pattern(constraints, id);
+                id
+            })
+            .collect_vec();
+
+        // Traverse the prefix tree from root to leaves and do two things:
+        // 1. Make the mutually exclusivity invariant hold:
+        //    at every node for any input, at most one constraint may be satisfied
+        // 2. Turn some of the states into deterministic transitions, according to
+        //    `make_det`
+
+        // With this `toposort` we are allowed to add vertices and edges
+        let mut traverser = self.matcher.toposort();
+        while let Some(state) = traverser.next(&self.matcher) {
+            // Make the mutually exclusivity invariant hold
+            self.make_mutex(state);
+            // Turn some of the states into deterministic transitions, according to
+            // `make_det`
+            let constraints = self.matcher.constraints(state).collect_vec();
+            if make_det(&constraints) {
```

Comment:
Could we have `make_det` take an `impl Iterator` instead of allocating multiple vec?
It's not too important, `make_mutex` already takes a lot of time.

Reply:
I would like that, but I'm not sure how to express that type without introducing a new trait. `make_det` would have to be generic over a `I: Iterator`, and that is not possible, right?